### PR TITLE
REP-6467 Puppet Master Upgrade

### PR DESCRIPTION
**Pull request by wdschei**
We went from v3.x to v6.x and fixed on the little things along the way.

### CODE REVIEW

File: Puppetfile [hunk 1]
```diff
@@ -3,37 +3,36 @@
...
+mod 'jamesnetherton-google_chrome', '0.3.0'
+mod 'garethr-scala', '0.1.2'
 mod 'garethr-docker', '5.3.0'
-mod 'jamesnetherton/google_chrome', '0.2.1'
-mod 'puppet-archive', '1.1.2'
 mod 'golja-influxdb', '4.0.0'
-mod 'puppet-grafana', '2.6.3'
-mod 'stankevich-python', '1.18.2'
+mod 'maestrodev-maven', '1.4.0'
+mod 'maestrodev-sonarqube', '2.6.7'
```
Comment (dmnjohns) [inline Puppetfile]: This might be one of those things that tricks Git because of when and how this line was changed. Could you rebase on master for me so that I can verify this line gets removed?

Comment (wdschei) [reply Puppetfile]: It was rebased just before opening the PR.

File: Puppetfile [hunk 2]
```diff
@@ -3,37 +3,36 @@
...
+mod 'mjanser-vagrant', '1.1.0'
```
Comment (dmnjohns) [inline Puppetfile]: It isn't really part of this story and would just be for cleanup, but I don't believe the following modules are used anymore:

- `mjanser-vagrant`
- `puppet-virtualbox`

File: Puppetfile (hunk 1 above)
Comment (wdschei) [reply Puppetfile]: I thought I got it during the conflict resolution, but I guess I didn't.
I'll manually remove it now.

Comment (dmnjohns) [reply Puppetfile]: Ah. In that case, this line should not be here.

Comment (wdschei) [reply Puppetfile]: Removed.

File: Puppetfile (hunk 2 above)
Comment (wdschei) [reply Puppetfile]: Removed.

File: bootstrap-puppet-master.sh [hunk 3]
```diff
@@ -1,57 +1,83 @@
 #!/bin/bash
 
-# set it up to work with debians now
-wget https://apt.puppetlabs.com/puppetlabs-release-wheezy.deb &&
-dpkg -i puppetlabs-release-wheezy.deb &&
-apt-get update &&
+# Applies to debian flavors only now.
+which lsb_release > /dev/null 2>&1
```
[the remark this answers is not in the document]
Comment (wdschei) [reply bootstrap-puppet-master.sh]: I have never used the `command` command before and have never been on an POSIX compliant OS where `which` didn't behave as expected.
However, more compliant is always better.
Updated.

[the remark this answers is not in the document]
Comment (wdschei) [reply bootstrap-puppet-master.sh]: I also updated the other uses of `which` too.

File: myModules/default_node/manifests/init.pp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+  class { 'firewall': }
+
+  # just setting swap size to a 4GB swapfile no matter what
+  # if this becomes a problem in the future, change it, or make it more dynamic
+  # the base::swap class will base it on the size of the ram, if not specified
+  base::swap { 'swapfile':
+          swapfile => '/swapfile',
+          swapsize => 4096,
+      }
```
[the remark this answers is not in the document]
Comment (wdschei) [reply myModules/default_node/manifests/init.pp]: Yes, it is.
I have been fighting IntelliJ on spacing in the `.pp` files this entire time.
I haven't been in them enough historically for it to be an issue, but this time has made it a bit painful.
Updated.

File: myModules/default_node/manifests/init.pp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  # have to actually include it!
+  class { 'repose_nagios::client': }
+}
```
[the remark this answers is not in the document]
Comment (wdschei) [reply myModules/default_node/manifests/init.pp]: Addeed.

File: myModules/repose_nexus/metadata.json [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
   "summary": "",
   "license": "Apache 2.0",
-  "source": "",
+  "source": "git://github.com/rackerlabs/repose-infrastructure-ng.git",
+  "project_page": "https://github.com/rackerlabs/repose-infrastructure-ng",
   "dependencies": [
     {"name":"puppetlabs-stdlib","version_requirement":">= 1.0.0"},
-    {"name":"repose/backup_cloud_files","version_requirement":">= 0.0.1"},
-    {"name":"maestrodev/wget","version_requirement":">= 0.0.1"}
+    {"name":"repose/backup_cloud_files","version_requirement":">= 0.0.1"}
```
Comment (dmnjohns) [inline myModules/repose_nexus/metadata.json]: Is there a reason this dependency wasn't updated to the dash-notation?

File: myModules/repose_nexus/metadata.json [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
+  "author": "repose",
   "summary": "",
   "license": "Apache 2.0",
-  "source": "",
+  "source": "git://github.com/rackerlabs/repose-infrastructure-ng.git",
+  "project_page": "https://github.com/rackerlabs/repose-infrastructure-ng",
   "dependencies": [
     {"name":"puppetlabs-stdlib","version_requirement":">= 1.0.0"},
-    {"name":"repose/backup_cloud_files","version_requirement":">= 0.0.1"},
-    {"name":"maestrodev/wget","version_requirement":">= 0.0.1"}
```
Comment (dmnjohns) [inline myModules/repose_nexus/metadata.json]: Is there a reason this dependency was dropped?

File: myModules/repose_nexus/metadata.json (hunk 6 above)
Comment (wdschei) [reply myModules/repose_nexus/metadata.json]: The `repose` ones are local and not published to Puppet Forge.

Comment (wdschei) [reply myModules/repose_nexus/metadata.json]: But they are going away now in favor of just using the `Puppetfile` and `librarian-puppet`.

Comment (dmnjohns) [reply myModules/repose_nexus/metadata.json]: Would the dash notation not work since it's a local module? I would think it still would, but I haven't tried it. The module name itself is in dash notation, however, so I strongly suspect it would work.

File: myModules/repose_nexus/metadata.json (hunk 7 above)
Comment (wdschei) [reply myModules/repose_nexus/metadata.json]: This particular box is not currently Puppet'd and that `wget` version caused conflicts with other module dependencies.

File: myModules/repose_nexus/metadata.json (hunk 6 above)
Comment (wdschei) [reply myModules/repose_nexus/metadata.json]: I don't know and didn't try. The other's I copied of the Forge.

File: myModules/repose_nexus/metadata.json (hunk 7 above)
Comment (dmnjohns) [reply myModules/repose_nexus/metadata.json]: Ah, that's right. I didn't really look at what box this was, but that's fair.

File: myModules/default_node/manifests/init.pp [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  #     purge => true,
+  # }
+
+  Firewall {
+      before  => Class['base::fw_post'],
+      require => Class['base::fw_pre'],
+  }
+
+  class { ['base::fw_pre', 'base::fw_post']: }
+  class { 'base': }
```
[the remark this answers is not in the document]
Comment (wdschei) [reply myModules/default_node/manifests/init.pp]: I took this back from the `Resource`-like behavior to the `Include`-like for consistency.
Updated.

File: puppet_apply/puppet_master.pp [hunk 9]
```diff
@@ -0,0 +1,2 @@
+include default_node
+include puppet_master
```
[the remark this answers is not in the document]
Comment (wdschei) [reply puppet_apply/puppet_master.pp]: This was an old file that was just moved, but for consistency with our current practices a newline was added.
Added.

File: myModules/puppet_master/manifests/init.pp [hunk 10]
```diff
@@ -1,102 +1,61 @@
...
+        require => Package["puppetserver"],
+    }
+
+    file{ '/etc/puppetlabs/puppet/eyaml':
+        owner  => puppet,
+        group  => puppet,
+        mode    => '0750',
+        require => Package["puppetserver"],
+        recurse => true,
+        purge => false,
```
[the remark this answers is not in the document]
Comment (wdschei) [reply myModules/puppet_master/manifests/init.pp]: I had to go look close to see when I changed this. LOL.
This was part of a global change to ensure all `mode`'s were in single ticks.
Formatting update though for consistency.
Updated.

File: myModules/repose_grafana/manifests/init.pp [hunk 11]
```diff
@@ -78,10 +78,12 @@
 
   #Papertrail the grafana logs
   $papertrail_port = hiera("base::papertrail_port", 1)
-  class { 'remotesyslog':
-    port => $papertrail_port,
-    logs => [
-      '/var/log/grafana/grafana.log'
+  class {'papertrail':
+    destination_host => 'logs.papertrailapp.com',
+    destination_port => 0 + $papertrail_port,
```
[the remark this answers is not in the document]
Comment (wdschei) [reply myModules/repose_grafana/manifests/init.pp]: We probably could and should when we update all of the hiera calls as our current use of them is deprecated.
However that was a little more invasive that I originally wanted to go with this initial migration.
Now that v6.3 is already out before we are done, maybe that should be part of that fast follow so our infrastructure doesn't rot again.

File: myModules/repose_phone_home/metadata.json [hunk 12]
```diff
@@ -7,7 +7,5 @@
   "source": "git://github.com/rackerlabs/repose-infrastructure-ng.git",
   "project_page": "https://github.com/rackerlabs/repose-infrastructure-ng",
   "dependencies": [
-    {"name":"puppetlabs-stdlib","version_requirement":">= 1.0.0"},
-    {"name":"puppetlabs-mongodb","version_requirement":">= 0.0.1"}
```
Comment (dmnjohns) [inline myModules/repose_phone_home/metadata.json]: I'm just picking a random instance of these removals to make this comment.

I am really not a fan of the idea of removing module-specific dependencies. I get that we want to minimize dependency conflicts, but removing dependencies from module metadata hard couples our modules to the overall project and loses information about which module(s) depend on which other module(s). Furthermore, listing our dependencies in the top-level `Puppetfile` does not prevent dependency conflicts in general, only for our own modules. That is, transitive dependencies could still conflict. And of course, Puppet advises module-authors to include dependencies in module metadata:
https://puppet.com/docs/puppet/6.3/bgtm.html#dependencies

I would personally suggest that modules keep their dependencies in their metadata. That way we still have complete modules that just happen to live in the same repository. Librarian Puppet can handle dependency resolution for us, just like it does for all of our transitive dependencies. Admittedly, we could end up in a situation where multiple modules that we have authored depend on the same module (in name) which is clearly not ideal since it could introduce conflict in the form of multiple sources of truth. However, we have control over our own modules and should notice that and be able to fix it readily enough.

In truth, I think the metadata dependencies approach more closely mirrors what we do in the [Repose project](https://github.com/rackerlabs/repose) with the major difference being that we can use the dependency recommender in that project to avoid the multiple sources of truth problem.

Actually, digging into this a bit more, I think there is a better solution still. Why are we creating complete modules? I can't think of a reason. We want to be able to group servers to apply common configuration, sure, but I believe there is a better way to do that. Enter the [Role and Profiles method](https://puppet.com/docs/pe/2018.1/the_roles_and_profiles_method.html) as detailed by Puppet Enterprise docs with an [accompanying example](https://github.com/puppetlabs/control-repo). This seems like what we are trying to do, just done better. If we structured this project as a control repo, it would make complete sense to manage all  of our dependencies in the `Puppetfile` (and sure, we could still use Librarian Puppet rather than r10k). That would get around my issue with these pseudo-modules.

**tl;dr:** This change is fine, but let's move closer to the roles and profiles design pattern at some point and stop treating our "module" code like modules.

File: myModules/repose_grafana/manifests/init.pp (hunk 11 above)
[the remark this answers is not in the document]
Comment (dmnjohns) [reply myModules/repose_grafana/manifests/init.pp]: Fair enough.

File: myModules/repose_phone_home/metadata.json (hunk 12 above)
Comment (wdschei) [reply myModules/repose_phone_home/metadata.json]: Just to officialize it, I concur with all of that and think updating how we actually use Puppet is a great next step once we have everything at least compatible with v6.x.

File: myModules/repose_nagios/manifests/server.pp [hunk 13]
```diff
@@ -11,6 +11,21 @@
...
+    file{ '/etc/nginx/conf.d/nagios_htpasswd':
+        ensure  => file,
+        mode    => '0640',
+        owner   => root,
+        group   => 'www-data',
+        require => [
+            Package['nginx']
+        ],
+        content => template('repose_nagios/htpasswd.erb'),
+        notify      => Service['nginx'],
```
Comment (dmnjohns) [inline myModules/repose_nagios/manifests/server.pp]: Weird spacing here.

File: myModules/repose_nagios/manifests/server.pp [hunk 14]
```diff
@@ -11,6 +11,21 @@
...
+        $nagios_admin_user => $nagios_admin_pass,
+    }
+    file{ '/etc/nginx/conf.d/nagios_htpasswd':
+        ensure  => file,
+        mode    => '0640',
+        owner   => root,
+        group   => 'www-data',
+        require => [
+            Package['nginx']
+        ],
```
Comment (dmnjohns) [inline myModules/repose_nagios/manifests/server.pp]: Shouldn't need to make this an `Array` since this is a single required resource. I don't think we have settled on a style for this since I see both approaches in our Puppets.

File: myModules/repose_nagios/templates/htpasswd.erb [hunk 15]
```diff
@@ -0,0 +1,4 @@
+# This file is managed by Puppet. DO NOT EDIT.
+<% @accounts.each do |user, password| -%>
+<%= user %>:<%= scope.call_function('apache_pw_hash', [password]) %>
+<% end -%>
```
Comment (wdschei) [inline myModules/repose_nagios/templates/htpasswd.erb]: Just to give credit where it is due, this is based directly on this blog post and accompanying gist:
 - https://dennis.dieploegers.de/a-simple-htpasswd-file-management-using-puppet/

File: myModules/repose_nagios/manifests/server.pp [hunk 16]
```diff
@@ -11,6 +11,21 @@
 
     include base::mail_sender
 
+    $accounts = {
+        $nagios_admin_user => $nagios_admin_pass,
+    }
+    file{ '/etc/nginx/conf.d/nagios_htpasswd':
```
Comment (dmnjohns) [inline myModules/repose_nagios/manifests/server.pp]: I like this approach better than the old approach. It's a shame that we lose salting, but given a strong password, not a huge loss. Especially considering what these credentials are for.
